src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,295 @@
+ * @author Kris Singh
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_GAN_IMPL_HPP
+#define MLPACK_METHODS_ANN_GAN_IMPL_HPP
+
+#include "gan.hpp"
+
+#include <mlpack/core.hpp>
+#include <boost/test/unit_test.hpp>
+#include <mlpack/tests/test_tools.hpp>
+
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/layer/base_layer.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/init_rules/network_init.hpp>
+#include <mlpack/methods/ann/visitor/output_parameter_visitor.hpp>
+
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/core/dists/gaussian_distribution.hpp>
+#include <mlpack/methods/ann/init_rules/random_init.hpp>
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::math;
+using namespace mlpack::distribution;
+
+namespace mlpack {
+namespace ann /** artifical neural network  */ {
+template<typename Model, typename InitializationRuleType>
+GAN<Model, InitializationRuleType>::GAN(
+    arma::mat& trainData,
+    arma::mat& noiseData,
+    Model& generator,
+    Model& discriminator,
+    InitializationRuleType initializeRule,
+    size_t batchSize,
+    size_t disIteration):
+    trainData(trainData),
+    noiseData(noiseData),
+    generator(generator),
+    discriminator(discriminator),
+    initializeRule(initializeRule),
+    trainGenerator(false),
+    batchSize(batchSize),
+    disIteration(2 * disIteration),
+    reset(false)
+{
+  // Insert IdentityLayer for Joing the generator and discriminator
+  discriminator.network.insert(
+      discriminator.network.begin(), 
+      new IdentityLayer<>());
+  trainGenerator = false;
+  counter = 1;
+  offset = 0;
+  iterationDiscriminator = disIteration;
+}
+
+template<typename Model, typename InitializationRuleType>
+void GAN<Model, InitializationRuleType>::Reset()
+{
+  size_t genWeights = 0;
+  size_t discWeights = 0;
+  NetworkInitialization<InitializationRuleType> networkInit(initializeRule);
+
+  for (size_t i = 0; i < generator.network.size(); ++i)
+    genWeights += boost::apply_visitor(weightSizeVisitor, generator.network[i]);
+
+  for (size_t i = 0; i < discriminator.network.size(); ++i)
+    discWeights += boost::apply_visitor(weightSizeVisitor, 
+        discriminator.network[i]);
+
+  parameter.set_size(genWeights + discWeights, 1);
+
+  // Intialise the parmaeters generator
+  networkInit.Initialize(generator.network, parameter);
+  // Intialise the parameters discriminator
+  networkInit.Initialize(discriminator.network, parameter, genWeights);
+  generator.Parameters() = arma::mat(
```

Comment:
Have you checked that the weights from the generator/discriminator are shared with the parameter matrix in the GAN class? The generator/discriminator should call the ResetParameters function on the first Evaluation call, since reset is false, so you probably end up with different weights.

What you could do here is:

```c++
generator.ResetParameters();
discriminator.ResetParameters();

size_t genWeights = generator.Parameters().n_elem;
size_t discWeights = discriminator.Parameters().n_elem;

parameter.set_size(generator.Parameters().n_elem +
    discriminator.Parameters().n_elem, 1);

NetworkInitialization<InitializationRuleType> networkInit(initializeRule);
networkInit.Initialize(generator.network, parameter);
networkInit.Initialize(discriminator.network, parameter, generator.Parameters().n_elem);
```